### DISCUSSION TITLE
neighborhood statistics

### DISCUSSION THREAD

**Discussion by davis-wesley-12345** [dtcenter/METplus#1929] [Statistical Computation] [answered]
I'm getting some discrepancies in the calculation of my neighborhood statistics between what comes out of MET and what came out of some known good code.

They're being calculated over the same area and with the same widths, but the numbers are different. 

Do you have any documentation on how exactly MET is calculating neighborhood stats?

I looked throughout the MET repository but I couldn't find anything, closest I got was here:

https://github.com/dtcenter/MET/tree/main_v10.1/met/src/libcode/

Any help is greatly appreciated, thank you!

**Comment by JohnHalleyGotway** [comment, marked as answer]
@davis-wesley-12345 I see you have a question about the neighborhood statistics computed by MET. And by that I'm guessing you mean the NBRCNT line written by the Grid-Stat tool.

The statistics are discussed in [this section](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=neighborhood#met-verification-measures-for-neighborhood-methods) of Appendix C in the User's Guide. And a description of the methodology can be found in [this section](https://met.readthedocs.io/en/latest/Users_Guide/grid-stat.html?highlight=neighborhood#neighborhood-methods) of the Grid-Stat chapter. But it sounds like you're looking for specific in the code instead of documentation.

So let me point you to some specifics. Grid-Stat loops over each `fcst.field` and `obs.field` comparison specified in the configuration file. For each, it loops over the options provided in the `nbrhd` dictionary as well as the categorical thresholds listed in the `cat_thresh` entry. For each combination, it calls the [fractional_coverage()](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/basic/vx_util/data_plane_util.cc#L207) function for the [forecast field](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/tools/core/grid_stat/grid_stat.cc#L1395) and the [obs field](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/tools/core/grid_stat/grid_stat.cc#L1435). And that function is a crucial step.

For each point in the grid, it replaces the value at that grid point with a number between 0 and 1 representing the fraction of grid points in the surrounding neighborhood that meet the specified threshold criteria. For points near the edge of the domain, where the neighborhood extends outside the domain, the `nbrhd.vld_thresh` config option tells it what to do. And we call the resulting field the "fractional coverage field".

When debugging things, I'd strongly recommend configuring Grid-Stat to set `nc_pairs_flag.nbrhd = TRUE;`. That'll write those fractional coverage fields to the NetCDF matched pairs file from Grid-Stat. And you can view those fields to look for unexpected results.

Note that there are several configuration options for the [nbrhd dictionary](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/data/config/GridStatConfig_default#L165) which could impact the results.
- `field = BOTH` computes the fractional coverage field for both the fcst and obs inputs. You'd only modify this default value if one of your inputs already has values between 0 and 1 that should be processed as a fractional coverage field.
- `shape = SQUARE` defines an NxN (where N is odd) square neighborhood around each grid point. Switching to CIRCLE instead lops off the corners of the neighborhood.
- `width` controls the size of the neighborhood as you'd expect. Output is produced for each width specified.

After the fractional coverage fields are computed, Grid-Stat picks out subsets of points based on the user-specified spatial masking regions, and for each subset it calls [do_nbrcnt()](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/tools/core/grid_stat/grid_stat.cc#L1555) to compute the neighborhood stats, including the Fractions Skill Score (FSS). And those NBRCNT stats are written to ascii output files.

Hope that helps. If you'd like to tell me more about the discrepancies you're seeing, I could look more closely.

**Reply by JohnHalleyGotway** [reply]
Just wanted to follow up on this discussion to see if questions remain. If not, please feel free to select the "Mark As Answer" button for one of the previous responses, and then I'll lock it to prevent future posts. If I don't hear otherwise in the next week or so, I'll just go ahead and mark it as being answered.

Thanks!
